core/arch/arm/mm/core_mmu.c

```diff
@@ -866,10 +913,12 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 			map->attr = core_mmu_type_to_attr(map->type);
 			va -= map->size;
 			va = ROUNDDOWN(va, map->region_size);
-#if !defined(CFG_WITH_LPAE)
-			/* Mapping does not yet support sharing L2 tables */
-			va = ROUNDDOWN(va, CORE_MMU_PGDIR_SIZE);
-#endif
+			/*
+			 * Make sure that va is aligned with pa for
+			 * efficient pgdir mapping. Basically pa &
+			 * pgdir_mask should be == va & pgdir_mask
+			 */
+			va += ((map->pa - va) & CORE_MMU_PGDIR_MASK);
```

Comment:
I think this will fail as this sequence assigns virtual addresses in decreasing order. Should rather be something like `va -= ((map->pa - va) & CORE_MMU_PGDIR_MASK);`, but please cross-check...

Reply:
Yeah, I thought about it. But `va = ROUNDDOWN(va, map->region_size);` at L913 moves `va` down. If I see it right, then `va` will be increased, but it will  not move over `map->region_size`.

I think, I'll try to simulate different cases to be sure.

Reply:
You can simply force the case by hacking this way:
```patch
--- a/core/arch/arm/mm/core_mmu.c
+++ b/core/arch/arm/mm/core_mmu.c
@@ -898,7 +898,7 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 
        assert(!((va | end) & SMALL_PAGE_MASK));
 
-       if (core_mmu_place_tee_ram_at_top(va)) {
+       if (1 || core_mmu_place_tee_ram_at_top(va)) {
                /* Map non-flat mapped addresses below flat mapped addresses */
                for (map = memory_map; !core_mmap_is_end_of_table(map); map++) {
                        if (map->va)
```

Reply:
Yep, I tried similar test on qemu (v7 and v8). It worked perfectly  fine, but now I'm worried that I might miss some corner case...